lib/Coffer/Path.hs

```diff
@@ -221,16 +224,31 @@ instance (Buildable path) => Buildable (QualifiedPath path) where
       Nothing -> build path
 
 instance (FromHttpApiData path) => FromHttpApiData (QualifiedPath path) where
-  parseUrlPiece text
-    | [pathPiece] <- T.splitOn "#" text = do
-        path <- parseUrlPiece @path pathPiece
-        pure $ QualifiedPath Nothing path
-    | [backendPiece, pathPiece] <- T.splitOn "#" text = do
-        backendName <- newBackendName backendPiece
-        path <- parseUrlPiece @path pathPiece
+  parseUrlPiece = mkQualifiedPath
+    (\_ -> "Invalid qualified path format. Expected [BACKENDNAME#]PATH")
+    parseUrlPiece
+
+mkQualifiedPath
+  :: (Text -> Text)
+  -> (Text -> Either Text path)
+  -> (Text -> Either Text (QualifiedPath path))
+mkQualifiedPath errorMessage mkPath text = case T.splitOn "#" text of
+      [backendNameStr, pathStr] -> do
+        backendName <- readBackendName' backendNameStr
+        path <- mkPath pathStr
         pure $ QualifiedPath (Just backendName) path
-    | otherwise =
-        Left "Invalid qualified path format. Expected [BACKENDNAME#]PATH"
+      [pathStr] -> do
+        path <- mkPath pathStr
+        pure $ QualifiedPath Nothing path
+      _ -> Left $ errorMessage text
```

Comment:
Instead of taking in an `errorMessage :: Text -> Text` parameter, let's just return an error message explaining what went wrong, like what the other `mk*` constructors do.

So, what went wrong in the `_` branch? Either `T.splitOn "#"` returned a list with 0 elements or a list with >2 elements.

I don't think `splitOn` ever returns an empty list, at least according to my experiments, so we can rule that out. That means it must have returned a list with >2 elements. Which means the string contained more than one '#' characters, so let's return an error message that says that.

Reply:
Done